tests/4.5/application_kernels/qmcpack_parallel_offload.c

```diff
@@ -0,0 +1,87 @@
+//===---- test_target_parallel.c - combined consutrct target and parallel  -------------===//
+// 
+// OpenMP API Version 4.5 Nov 2015
+// 
+// This test checks for multiple parallel offloading kernels. 
+// Kernel1 does parallel offload without specifying a device id while 
+// kernel2 tries to offload to specific devices.
+// Adapted from https://reviews.llvm.org/D74145 to match V&V coding and
+// reporting style. 
+//
+//===----------------------------------------------------------------------------------===//
+//
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <time.h>
+
+#include "ompvv.h"
+
+
+void qmcpack_kernel1(int ndev) {
+
+  const int N = 1024, num_teams = 8, num_th = 4;
+  OMPVV_INFOMSG("qmcpack_multiple_threads_offloading");
+
+  //Spawn 1 thread per device
+#pragma omp parallel for num_threads(ndev)
+  for (int i = 0; i < ndev; ++i) {
+#pragma omp target teams distribute parallel for num_teams(num_teams) num_threads(num_th)
+    for (size_t i = 0; i < N; ++i) {
+      for (size_t j = 0; j < N/2 ; ++j) {
+        int tmp_val = N *i*j;
+      }
+    }
+  }
+}
+void qmcpack_kernel2(int ndev) {
+
+  const int N = 1024, num_teams = 8, num_th = 4;
+  OMPVV_INFOMSG("qmcpack_multiple_threads_offloading");
+
+  //Spawn 1 thread per device
+#pragma omp parallel for num_threads(ndev)
+  for (int i = 0; i < ndev; ++i) {
+#pragma omp target teams distribute parallel for device(i % ndev) num_teams(num_teams) num_threads(num_th)
+    for (size_t i = 0; i < N; ++i) {
+      for (size_t j = 0; j < N/2 ; ++j) {
+        int tmp_val = N *i*j;
+      }
+    }
+  }
+}
+
+int main(int argc, char *argv[]) {
+  int i;
+  const int N = 1000;
+  int num_dev = omp_get_num_devices();
+  //printf("No. of devices available %d.\n", num_dev);
+  
+  OMPVV_WARNING_IF(num_dev <= 1, "Multiple devices not available");
+
+  OMPVV_TEST_OFFLOADING;
+
+  clock_t start = clock();
+
+  for (int i = 0; i < N; ++i) {
+    qmcpack_kernel1(num_dev);
+  }
+
+  const clock_t duration1 = (clock() - start) * 1000 / CLOCKS_PER_SEC / N;
+  
+  if(num_dev > 1){
+    start = clock();
+    for (int i = 0; i < N; ++i) {
+      qmcpack_kernel2(num_dev);
+    }
+    const clock_t duration2 = (clock() - start) * 1000 / CLOCKS_PER_SEC / N;
+    printf("Avg time for parallel offloading to multiple devices: %ld ms\n",duration2);
```

Comment:
I suggest we use OMPVV_INFOMSG

Reply:
Will do that after the test is finalized.

Reply:
Actually, I don't know if it appropriate as I want the timings to be printed without additional flags.

Reply:
The idea of the TEST_VERBOSE flag was to be able to obtain a simple pass fail report and no other output. If that's the case, then we either want to re-evaluate this design in the infrastructure, or provide a new flag. I don't think it is a good idea to have `printf`s with no other information.

Reply:
This test is a little different from our usual conformance tests. If we see more of these then maybe we should think of a new macro.

Reply:
I agree this is a different to our conformance tests. That's why I am conflicted about adding information that is output in this test (and only in this test) regarding performance, when our focus has always been V&V. If we want to consider performance, we should consider how to change the infrastructure for performance.

Reply:
V nice to see an app. test code! I like that we are printing timings!! Probably better to stick with the macros/ infrastructure design/software practice by turning on TEST_VERBOSE and using OMPVV_INFOMSG - that way we are strictly following our rules that we have set and then encourage anyone else to do the same

Reply:
INFOMSGs were designed to give additional information for tests, I fear that the timings will get buried in all other messages. That said, current format is for easy testing during discussion phase where they may want to take it for a test drive. I am all in to design a macro that better fits timing reporting.

Reply:
Yes. This is definitely needed. So far we had considered giving timing information for the duration of the whole test, but we have not discussed what happens when we want to give intermediate timing results.